vdslib/src/main/java/com/yahoo/vdslib/state/ClusterState.java

```diff
@@ -18,25 +19,183 @@
 
     private static final NodeState DEFAULT_STORAGE_UP_NODE_STATE = new NodeState(NodeType.STORAGE, State.UP);
     private static final NodeState DEFAULT_DISTRIBUTOR_UP_NODE_STATE = new NodeState(NodeType.DISTRIBUTOR, State.UP);
+    private static final NodeState DEFAULT_STORAGE_DOWN_NODE_STATE = new NodeState(NodeType.STORAGE, State.DOWN);
+    private static final NodeState DEFAULT_DISTRIBUTOR_DOWN_NODE_STATE = new NodeState(NodeType.DISTRIBUTOR, State.DOWN);
 
-    private int version = 0;
-    private State state = State.DOWN;
-    // nodeStates maps each of the non-up nodes that have an index <= the node count for its type.
-    private Map<Node, NodeState> nodeStates = new TreeMap<>();
+    /**
+     * Maintains a bitset where all non-down dows have a bit set. All nodes that differs from defaultUp
+     * and defaultDown are store explicit in a hash map.
+     */
+    private static class Nodes {
+        private int maxIndex;
+        private final NodeType type;
+        private final BitSet upNodes;
+        private final Map<Integer, NodeState> nodeStates = new HashMap<>();
+        Nodes(NodeType type) {
+            this.type = type;
+            upNodes = new BitSet();
+        }
+        Nodes(Nodes b) {
+            maxIndex = b.maxIndex;
+            type = b.type;
+            upNodes = (BitSet) b.upNodes.clone();
+            b.nodeStates.forEach((key, value) -> nodeStates.put(key, value.clone()));
+        }
+
+        void updateMaxIndex(int index) {
+            if (index > maxIndex) {
+                upNodes.set(maxIndex, index);
+                maxIndex = index;
+            }
+        }
+
+        int getMaxIndex() { return maxIndex; }
+
+        NodeState getNodeState(int index) {
+            NodeState ns = nodeStates.get(index);
+            if (ns != null) return ns;
+            return (index >= getMaxIndex() || ! upNodes.get(index))
+                    ? new NodeState(type, State.DOWN)
+                    : new NodeState(type, State.UP);
```

Comment:
More fuel for the "node states should be immutable" fire... 🔥

Reply:
This breaks down if someone does getNodeState(node_that_is_up).modify_some_important_field.